File: src/src_platform_swing/fmg/swing/res/img/Logo.java
package fmg.swing.res.img;

import java.awt.*;
import java.awt.geom.Point2D;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.swing.Icon;

import fmg.common.geom.PointDouble;
import fmg.swing.Cast;
import fmg.swing.utils.ImgUtils;

/** main logos image */
public class Logo extends fmg.core.img.Logo<Icon> implements Icon {

   public Logo(boolean useGradient) {
      super(useGradient);
   }

   @Override
   public int getIconWidth() { return (int)getWidth(); }

   @Override
   public int getIconHeight() { return (int)getHeight(); }

   @Override
   protected Icon createImage() {
      BufferedImage img = new BufferedImage(getIconWidth(), getIconHeight(), BufferedImage.TYPE_INT_ARGB);
      Graphics2D g = img.createGraphics();
      g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
      g.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY);
      draw(g);
      g.dispose();
      return ImgUtils.toIco(img);
   }

   @Override
   protected void drawImage(Icon img) {
      // none... already drawed in createImage()
   }

   private void draw(Graphics2D g) {
      final int iPenWidth = 2;

      List<PointDouble> rays0 = new ArrayList<>();
      List<PointDouble> inn0 = new ArrayList<>();
      List<PointDouble> oct0 = new ArrayList<>();
      getCoords(rays0, inn0, oct0);

      Point2D.Double [] rays = rays0.stream().map(p -> Cast.toPoint(p)).toArray(size -> new Point2D.Double[size]);
      Point2D.Double [] inn  = inn0 .stream().map(p -> Cast.toPoint(p)).toArray(size -> new Point2D.Double[size]);
      Point2D.Double [] oct  = oct0 .stream().map(p -> Cast.toPoint(p)).toArray(size -> new Point2D.Double[size]);
      Point2D.Double center = new Point2D.Double(getWidth()/2, getHeight()/2);

      Color [] palette = Arrays.stream(Palette).map(clr -> Cast.toColor(clr)).toArray(size -> new Color[size]);

      // paint owner gradient rays
      for (int i=0; i<8; i++) {
         if (isUseGradient()) {
            // rectangle gragient
            g.setPaint(new GradientPaint(oct[(i+5)%8], palette[(i+0)%8], oct[i], palette[(i+3)%8]));
            g.fillPolygon(new int[] {
                  (int)rays[i].x,
                  (int)oct[i].x,
                  (int)inn[i].x,
                  (int)oct[(i+5)%8].x
               }, new int[] {
                  (int)rays[i].y,
                  (int)oct[i].y,
                  (int)inn[i].y,
                  (int)oct[(i+5)%8].y
               }, 4);

            // emulate triangle gradient (see BmpLogo.cpp C++ source code)
            Color clr = palette[(i+6)%8];
            clr = new Color(clr.getRed(), clr.getGreen(), clr.getBlue(), 0);
            g.setPaint(new GradientPaint(center, clr, inn[(i+6)%8], palette[(i+3)%8]));
            g.fillPolygon(new int[] {
                  (int)rays[i].x,
                  (int)oct[i].x,
                  (int)inn[i].x
               }, new int[] {
                  (int)rays[i].y,
                  (int)oct[i].y,
                  (int)inn[i].y
               }, 3);
            g.setPaint(new GradientPaint(center, clr, inn[(i+2)%8], palette[(i+0)%8]));
            g.fillPolygon(new int[] {
                  (int)rays[i].x,
                  (int)oct[(i+5)%8].x,
                  (int)inn[i].x
               }, new int[] {
                  (int)rays[i].y,
                  (int)oct[(i+5)%8].y,
                  (int)inn[i].y
               }, 3);
         } else {
            g.setColor(Cast.toColor(Palette[i].darker()));
            g.fillPolygon(new int [] {
                  (int)rays[i].x,
                  (int)oct[i].x,
                  (int)inn[i].x,
                  (int)oct[(i+5)%8].x
               }, new int [] {
                  (int)rays[i].y,
                  (int)oct[i].y,
                  (int)inn[i].y,
                  (int)oct[(i+5)%8].y
               }, 4);
         }
      }

      // paint star perimeter
      g.setStroke(new BasicStroke(iPenWidth));
      for (int i=0; i<8; i++) {
         Point2D.Double p1 = rays[(i + 7)%8];
         Point2D.Double p2 = rays[i];
         g.setColor(palette[i]);
         g.drawLine((int)p1.x, (int)p1.y, (int)p2.x, (int)p2.y);
      }

      // paint inner gradient triangles
      for (int i=0; i<8; i++) {
         if (isUseGradient())
            g.setPaint(new GradientPaint(
                  inn[i], palette[(i+6)%8],
                  center, ((i&1)==0) ? Color.BLACK : Color.WHITE));
         else
            g.setColor(((i & 1) == 0)
                  ? Cast.toColor(Palette[(i + 6)%8].brighter())
                  : Cast.toColor(Palette[(i + 6)%8].darker()));
         g.fillPolygon(new int [] {
               (int)inn[(i + 0)%8].x,
               (int)inn[(i + 3)%8].x,
               (int)center.x
            }, new int [] {
               (int)inn[(i + 0)%8].y,
               (int)inn[(i + 3)%8].y,
               (int)center.y
            }, 3);
      }
   }

   @Override
   public void paintIcon(Component c, Graphics g, int x, int y) {
      getImage().paintIcon(c, g, x, y);
   }

   public static void main(String[] args) {
      TestDrawing.testApp2(size -> ImgUtils.zoom(new Logo(true), size, size));
   }

}

File: src/src_platform_uwp/fmg/uwp/res/img/win2d/Logo.cs
using System.Collections.Generic;
using Windows.UI.Core;
using Windows.UI.Xaml;
using Windows.Graphics.Display;
using Microsoft.Graphics.Canvas;
using fmg.common;
using fmg.common.geom;
using fmg.uwp.utils;
using fmg.uwp.draw.mosaic.win2d;
using fmg.core.img;

namespace fmg.uwp.res.img.win2d {

   /// <summary> Main logos image
   /// <br/>
   /// Win2D impl
   /// </summary>
   public abstract class Logo<TImage> : ALogo<TImage>
      where TImage : DependencyObject, ICanvasResourceCreator
   {

      static Logo() {
         if (StaticImgConsts.DeferrInvoker == null)
            StaticImgConsts.DeferrInvoker = doRun => AsyncRunner.InvokeFromUiLater(() => doRun(), CoreDispatcherPriority.Normal);
         if (RotatedImgConst.TimerCreator == null)
            RotatedImgConst.TimerCreator = () => new ui.Timer();
      }

      protected Logo() {
         SyncDraw = Windows.ApplicationModel.DesignMode.DesignModeEnabled;
      }

      protected void DrawBody(CanvasDrawingSession ds, bool fillBk) {
         ICanvasResourceCreator rc = Image;

         if (fillBk) {
            var bkClr = BackgroundColor;
            if (bkClr.A != Color.Transparent.A)
               ds.Clear(bkClr.ToWinColor());
         }

         IList<PointDouble> rays = new List<PointDouble>(), inn = new List<PointDouble>(), oct = new List<PointDouble>();
         GetCoords(rays, inn, oct);

         var center = new PointDouble(Width / 2.0, Height / 2.0);

         // paint owner rays
         for (var i = 0; i < 8; i++) {
            using (var geom = rc.BuildGeom(rays[i], oct[i], inn[i], oct[(i + 5) % 8])) {
               if (UseGradient) {
                  // rectangle gragient
                  using (var br = rc.CreateGradientPaintBrush(oct[(i + 5) % 8], Palette[(i + 0) % 8].ToColor(), oct[i], Palette[(i + 3) % 8].ToColor())) {
                     ds.FillGeometry(geom, br);
                  }

                  // emulate triangle gradient (see BmpLogo.cpp C++ source code)
                  var clr = Palette[(i + 6) % 8].ToColor();
                  clr.A = 0;
                  using (var br = rc.CreateGradientPaintBrush(center, clr, inn[(i + 6) % 8], Palette[(i + 3) % 8].ToColor())) {
                     using (var geom2 = rc.BuildGeom(rays[i], oct[i], inn[i])) {
                        ds.FillGeometry(geom2, br);
                     }
                  }
                  using (var br = rc.CreateGradientPaintBrush(center, clr, inn[(i + 2) % 8], Palette[(i + 0) % 8].ToColor())) {
                     using (var geom2 = rc.BuildGeom(rays[i], oct[(i + 5) % 8], inn[i])) {
                        ds.FillGeometry(geom2, br);
                     }
                  }
               } else {
                  ds.FillGeometry(geom, Palette[i].ToColor().Darker().ToWinColor());
               }
            }
         }

         // paint star perimeter
         var zoomAverage = (ZoomX + ZoomY) / 2;
         var penWidth = (float)(2*zoomAverage);
         for (var i = 0; i < 8; i++) {
            var p1 = rays[(i + 7) % 8];
            var p2 = rays[i];
            ds.DrawLine(p1.ToVector2(), p2.ToVector2(), Palette[i].ToColor().ToWinColor(), penWidth);
         }

         // paint inner gradient triangles
         for (var i = 0; i < 8; i++) {
            using (var geom = rc.BuildGeom(inn[(i + 0) % 8], inn[(i + 3) % 8], center)) {
               if (UseGradient) {
                  using (var br = rc.CreateGradientPaintBrush(inn[i], Palette[(i + 6) % 8].ToColor(), center, ((i & 1) == 0) ? Color.Black : Color.White)) {
                     ds.FillGeometry(geom, br);
                  }
               } else {
                  ds.FillGeometry(geom, ((i & 1) == 0)
                        ? Palette[(i + 6) % 8].ToColor().Brighter().ToWinColor()
                        : Palette[(i + 6) % 8].ToColor().Darker().ToWinColor());
               }
            }
         }
      }

      /////////////////////////////////////////////////////////////////////////////////////////////////////
      //    custom implementations
      /////////////////////////////////////////////////////////////////////////////////////////////////////

      /// <summary> Main logos image
      /// <br/>
      /// CanvasBitmap impl
      /// </summary>
      public class CanvasBmp : Logo<CanvasBitmap> {

         private readonly ICanvasResourceCreator _rc;

         public CanvasBmp(ICanvasResourceCreator resourceCreator) {
            _rc = resourceCreator;
         }

         protected override CanvasBitmap CreateImage() {
            var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
            return new CanvasRenderTarget(_rc, Width, Height, dpi);
         }

         protected override void DrawBody() {
            using (var ds = ((CanvasRenderTarget)Image).CreateDrawingSession()) {
               DrawBody(ds, true);
            }
         }
      }

   }

}
